### PR TITLE
Add account transactions table.

### PR DESCRIPTION
**Pull request by Benzbeeb**
## Implementation details
1. Add account transactions table.
2. Create a data structure for tracking the accounts that need balance update at the end of the block and the accounts related to the current processing transaction.
3. Insert set account msg after new block message then flush the messages to Kafka
---

Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: chain/emitter/app.go [hunk 1]
```diff
@@ -30,9 +31,10 @@ type App struct {
 	// Main Kafka writer instance.
 	writer *kafka.Writer
 	// Temporary variables that are reset on every block.
-	txIdx int              // The current transaction's index on the current block starting from 1.
-	accs  []sdk.AccAddress // The accounts that need balance update at the end of block.
-	msgs  []Message        // The list of all messages to publish for this block.
+	txIdx       int             // The current transaction's index on the current block starting from 1.
+	accsInBlock map[string]bool // The accounts that need balance update at the end of block.
+	accsInTx    map[string]bool
```
Comment (sorawit) [inline chain/emitter/app.go]: The accounts related to the current processing transaction.

File: chain/emitter/app.go [hunk 2]
```diff
@@ -202,22 +214,31 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
...
+	accMap := make(map[string]bool)
+
+	for acc, _ := range app.accsInTx {
+		accsInTx = append(accsInTx, acc)
+		accMap[acc] = true
+	}
+
+	txDict["account_transcations"] = accsInTx
+
+	fmt.Println("!!", accsInTx, "!!")
```
Comment (sorawit) [inline chain/emitter/app.go]: ?

File: flusher/flusher/handler.py [hunk 3]
```diff
@@ -26,7 +27,18 @@ def handle_new_block(self, msg):
...
 
     def handle_new_transaction(self, msg):
-        self.conn.execute(transactions.insert(), msg)
+        print(msg)
+        related_tx_accounts = msg["account_transcations"]
+        del msg["account_transcations"]
+        res = self.conn.execute(transactions.insert(), msg)
+        tx_id = res.inserted_primary_key[0]
+        print(related_tx_accounts)
+        print(tx_id)
```
Comment (sorawit) [inline flusher/flusher/handler.py]: ???

File: flusher/flusher/handler.py [hunk 4]
```diff
@@ -26,7 +27,18 @@ def handle_new_block(self, msg):
...
+        for account in related_tx_accounts:
+            print({"id": tx_id, "address": account})
+            if account != "":
```
Comment (sorawit) [inline flusher/flusher/handler.py]: This if should not be necessary

File: chain/emitter/distribution.go [hunk 5]
```diff
@@ -35,6 +35,13 @@ func (app *App) emitUpdateValidatorReward(addr sdk.ValAddress) {
...
-	// TODO: Add the beneficiary to related accounts
+	app.AddAccountsInTx(msg.DelegatorAddress.String())
 	app.emitUpdateValidatorReward(msg.ValidatorAddress)
 }
+
+// handleMsgSetWithdrawAddress implements emitter handler for MsgSetWithdrawAddress.
+func (app *App) handleMsgSetWithdrawAddress(
+	txHash []byte, msg dist.MsgSetWithdrawAddress, evMap EvMap, extra JsDict,
+) {
+	app.AddAccountsInTx(msg.WithdrawAddress.String(), msg.DelegatorAddress.String())
```
Comment (sorawit) [inline chain/emitter/distribution.go]: remove delegator address?

File: chain/emitter/distribution.go [hunk 6]
```diff
@@ -35,6 +35,13 @@ func (app *App) emitUpdateValidatorReward(addr sdk.ValAddress) {
 func (app *App) handleMsgWithdrawDelegatorReward(
 	txHash []byte, msg dist.MsgWithdrawDelegatorReward, evMap EvMap, extra JsDict,
 ) {
-	// TODO: Add the beneficiary to related accounts
+	app.AddAccountsInTx(msg.DelegatorAddress.String())
```
Comment (sorawit) [inline chain/emitter/distribution.go]: don't think this is right.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 7]
```diff
@@ -202,22 +213,29 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
 			"extra": extra,
 		})
 	}
+
+	accsInTx := []string{}
+	accMap := make(map[string]bool)
```
Comment (taobun) [inline chain/emitter/app.go]: Why we need to declare this map, we can't use app.AddAccountsInBlock(app.accsInTx...)?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 8]
```diff
@@ -61,9 +62,18 @@ func NewBandAppWithEmitter(
 	}
 }
 
-// AddAccounts adds the given accounts to the list of accounts to update balances end-of-block.
-func (app *App) AddAccounts(acc ...sdk.AccAddress) {
-	app.accs = append(app.accs, acc...)
+// AddAccountsBlock adds the given accounts to the list of accounts to update balances end-of-block.
+func (app *App) AddAccountsInBlock(accs ...string) {
```
Comment (sorawit) [inline chain/emitter/app.go]: This should take AccAddress. Ditto for below.

File: chain/emitter/app.go [hunk 9]
```diff
@@ -200,32 +211,40 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
...
+	for acc, _ := range app.accsInTx {
+		accsInTx = append(accsInTx, acc)
+	}
+
+	txDict["account_transcations"] = accsInTx
+	app.AddAccountsInBlock(accsInTx...)
+	app.accsInTx = make(map[string]bool)
 	txDict["messages"] = messages
-	app.AddAccounts(stdTx.GetSigners()[0])
+	app.AddAccountsInBlock(stdTx.GetSigners()[0].String())
```
Comment (sorawit) [inline chain/emitter/app.go]: Signers also need to get added to "account_transactions"

File: chain/emitter/app.go [hunk 10]
```diff
@@ -200,32 +211,40 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
...
 		})
 	}
+
+	accsInTx := []string{}
+
+	for acc, _ := range app.accsInTx {
+		accsInTx = append(accsInTx, acc)
+	}
+
+	txDict["account_transcations"] = accsInTx
```
Comment (sorawit) [inline chain/emitter/app.go]: Let's call this "related_accounts". Make sure to also update flusher.

File: chain/emitter/app.go [hunk 11]
```diff
@@ -200,32 +211,40 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
 			"extra": extra,
 		})
 	}
+
+	accsInTx := []string{}
+
```
Comment (sorawit) [inline chain/emitter/app.go]: nit: Remove extra space

File: chain/emitter/distribution.go [hunk 12]
```diff
@@ -35,6 +35,14 @@ func (app *App) emitUpdateValidatorReward(addr sdk.ValAddress) {
 func (app *App) handleMsgWithdrawDelegatorReward(
 	txHash []byte, msg dist.MsgWithdrawDelegatorReward, evMap EvMap, extra JsDict,
 ) {
-	// TODO: Add the beneficiary to related accounts
+	delegatorAddr := app.DistrKeeper.GetDelegatorWithdrawAddr(app.DeliverContext, msg.DelegatorAddress)
```
Comment (sorawit) [inline chain/emitter/distribution.go]: This should be called `withdrawAddr`

File: chain/emitter/bank.go [hunk 13]
```diff
@@ -5,16 +5,20 @@ import (
 )
 
 // handleMsgSend implements emitter handler for MsgSend.
-func (app *App) handleMsgSend(msg bank.MsgSend) {
-	app.AddAccounts(msg.FromAddress, msg.ToAddress)
+func (app *App) handleMsgSend(
+	txHash []byte, msg bank.MsgSend, evMap EvMap, extra JsDict,
+) {
+	app.AddAccountsInTx(msg.FromAddress.String(), msg.ToAddress.String())
```
Comment (sorawit) [inline chain/emitter/bank.go]: From is not necessary as it's already the signer.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 14]
```diff
@@ -200,32 +211,39 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
 			"extra": extra,
 		})
 	}
+	app.AddAccountsInTx(stdTx.GetSigners()[0])
```
Comment (sorawit) [inline chain/emitter/app.go]: Let's append all the signers here.

File: chain/emitter/app.go [hunk 15]
```diff
@@ -200,32 +211,39 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
...
+	for accStr, _ := range app.accsInBlock {
+		acc, _ := sdk.AccAddressFromBech32(accStr)
+		msgs = append(msgs, Message{
+			Key: "SET_ACCOUNT",
+			Value: JsDict{
+				"address": acc,
+				"balance": app.BankKeeper.GetCoins(app.DeliverContext, acc).String(),
+			}})
+	}
+	// Index 0 is message new block, We insert set account message after new block message
```
Comment (sorawit) [inline chain/emitter/app.go]: // Index 0 is message **NEW_BLOCK**, **w**e insert **SET_ACCOUNT** message**s** right after it.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: flusher/flusher/db.py [hunk 16]
```diff
@@ -244,3 +245,10 @@ def Column(*args, **kwargs):
     Column("completion_time", CustomDateTime),
     Column("amount", sa.DECIMAL),
 )
+
+account_transcations = sa.Table(
+    "account_transcations",
+    metadata,
+    Column("id", sa.Integer, sa.ForeignKey("transactions.id"), primary_key=True),
```
Comment (sorawit) [inline flusher/flusher/db.py]: Call this `transaction_id`. We will have `account_id` soon.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 17]
```diff
@@ -32,9 +32,10 @@ type App struct {
 	// Main Kafka writer instance.
 	writer *kafka.Writer
 	// Temporary variables that are reset on every block.
-	txIdx int              // The current transaction's index on the current block starting from 1.
-	accs  []sdk.AccAddress // The accounts that need balance update at the end of block.
-	msgs  []Message        // The list of all messages to publish for this block.
+	txIdx       int             // The current transaction's index on the current block starting from 1.
```
Comment (sorawit) [inline chain/emitter/app.go]: Let's remove this.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/emitter/handler.go [hunk 18]
```diff
@@ -40,26 +40,32 @@ func (app *App) handleMsg(txHash []byte, msg sdk.Msg, log sdk.ABCIMessageLog, ex
...
 	case oracle.MsgActivate:
 		app.handleMsgActivate(txHash, msg, evMap, extra)
 	case staking.MsgCreateValidator:
-		app.handleMsgCreateValidator(msg)
+		app.handleMsgCreateValidator(txHash, msg, evMap, extra)
 	case staking.MsgEditValidator:
-		app.handleMsgEditValidator(msg)
+		app.handleMsgEditValidator(txHash, msg, evMap, extra)
 	case staking.MsgDelegate:
 		app.handleMsgDelegate(msg)
```
Comment (taobun) [inline chain/emitter/handler.go]: Change function signature

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 19]
```diff
@@ -200,32 +208,38 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
 			"extra": extra,
 		})
 	}
+	app.AddAccountsInTx(stdTx.GetSigners()...)
+	accsInTx := []sdk.AccAddress{}
```
Comment (sorawit) [inline chain/emitter/app.go]: Let's call this `relatedAccounts` and initalize it with capacility of `len(app.accsInTx)`

File: chain/emitter/app.go [hunk 20]
```diff
@@ -200,32 +208,38 @@ func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
...
-		})
-	}
-
 	for _, event := range res.Events {
 		app.handleBeginBlockEndBlockEvent(event)
 	}
-
+	// Update balances of all affected accounts on this block.
+	// Index 0 is message NEW_BLOCK, we insert SET_ACCOUNT messages right after it.
+	modifyMsgs := []Message{app.msgs[0]}
```
Comment (sorawit) [inline chain/emitter/app.go]: modified

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**